### PR TITLE
feat(coordinator): add crypto and proof generation services

### PR DESCRIPTION
**Pull request by 0xmad**
# Description

- [x] Add key generation script
- [x] Add crypto service
- [x] Add proof generation service
- [x] Add launch instructions and code docs

## Additional Notes

Auth-z, controller and e2e tests are not implemented yet, but will be added in next PRs.
Current PR has atomic change for maci-coordinator package and doesn't affect anything.

## Related issue(s)

Related to https://github.com/privacy-scaling-explorations/maci-rpgf/issues/60

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: coordinator/.env.example [hunk 1]
```diff
@@ -1,2 +1,22 @@
+# Rate limit configuation
 TTL=60000
 LIMIT=10
+
+# Make sure your private and public RSA keys are generated (see package.json scripts)
+# Public key must be copied to frontend app to encrypt user sensitive data
```
Comment (ctrlc03) [inline coordinator/.env.example:6]: what user data? are you referring to coordinator's key?

File: coordinator/.env.example [hunk 2]
```diff
@@ -1,2 +1,22 @@
...
+COORDINATOR_PUBLIC_KEY_PATH="./pub.key"
+COORDINATOR_PRIVATE_KEY_PATH="./priv.key"
+
+# Make sure you have TallyVotesQv.zkey, TallyVotesQv.wasm
+# and TallyVotesNonQv.zkey, TallyVotesNonQv.wasm inside COORDINATOR_ZKEY_PATH folder
+# https://maci.pse.dev/docs/trusted-setup
+COORDINATOR_TALLY_ZKEY_NAME=TallyVotes
+
+# Make sure you have ProcessMessagesQv.zkey, ProcessMessagesQv.wasm 
+# and ProcessMessagesNonQv.zkey, ProcessMessagesNonQv.wasm inside COORDINATOR_ZKEY_PATH folder
```
Comment (ctrlc03) [inline coordinator/.env.example:16]: I think we should consider adding variables related to `rapidsnark` and c++ witness gen. Running rounds with the WASM witness gens is gonna take 100x more times to generate proofs

File: coordinator/ts/proof/proof.service.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      maciContractAddress,
+      tallyContractAddress,
+      tally: this.getZkeyFiles(process.env.COORDINATOR_TALLY_ZKEY_NAME!, useQuadraticVoting),
+      mp: this.getZkeyFiles(process.env.COORDINATOR_MESSAGE_PROCESS_ZKEY_NAME!, useQuadraticVoting),
+      outputDir: path.resolve("./proofs"),
+      tallyOutputFile: path.resolve("./tally"),
+      useQuadraticVoting,
+    });
+
+    const [processProofs, tallyProofs] = await Promise.all([
```
Comment (ctrlc03) [inline coordinator/ts/proof/proof.service.ts:117]: I'm afraid these need to be sequential, can't tally until all messages have been processed. Though we could run full local message processing (in batches as normal), with concurrent proof generation for it (as each batch is ready); and once the messages are processed locally (proofs might not have finished yet) in `Poll`, we can start tallying in `Poll` + as batches ready, start proof generation for tally

**Review by ctrlc03 — COMMENTED** [review]
just leaving some comments in the meantime that this is ready, great work so far though

File: coordinator/.env.example (hunk 1 above)
Comment (0xmad) [reply coordinator/.env.example:6]: Right, for now it's only coordinator private key.

File: coordinator/ts/proof/proof.service.ts (hunk 3 above)
Comment (0xmad) [reply coordinator/ts/proof/proof.service.ts]: Proof parallel processing will be implemented later, but good to know that I need to process them sequentially.

**Review by ctrlc03 — APPROVED** [review]
Thanks!

**Pull request merged**